sycl/doc/SYCL_compiler_and_runtime_design.md

```diff
@@ -0,0 +1,377 @@
+
+Ahead-of-time compilation is the process of invoking the back-end at compile
+time to produce the final binary, as opposed to just-in-time (JIT) compilation
+when final code generation is deferred until application runtime time.
+
+AOT compilation reduces application execution time by skipping JIT compilation
+and final device code can be tested before deploy.
+
+JIT compilation provides portability of device code and target specific
+optimizations.
+
+#### List of native targets
+
+The ahead-of-time compilation mode implies that there must be a way to specify
+a set of target architectures for which to compile device code. By default the
+compiler generates SPIR-V and OpenCL device JIT compiler produces native target
+binary.
+
+There are existing options for OpenMP* offload:
+
+`-fopenmp-targets=triple1,triple2`
+
+would produce binaries for target architectures identified by target triples
+`triple1` and `triple2`.
+
+A similar approach is used for SYCL:
+
+`-fsycl-targets=triple1,triple2`
+
+will produce binaries from SYCL kernels for devices identified by the two
+target triples. This basically tells the driver which device compilers must be
+invoked to compile the SYCL kernel code. By default, the JIT compilation
+approach is assumed and device code is compiled for a single target triple -
+`[spir,spir64]-\*-\*`.
+
+#### Device code formats
+
+Each target may support a number of code forms, each device compiler defines
+and understands mnemonics designating a particular code form, for example
+"`visa:3.3`" could designate virtual ISA version 3.3 for Intel GPU target (Gen
+architecture).  User can specify desired code format using the target-specific
+option mechanism, similar to OpenMP.
+
+`-Xsycl-target=<triple> <arg>`
+
+For example, to support offload to CPU, FPGA, Gen9/vISA3.3, Gen9/SPIR-V the
+following options would be used:
+
+`-fsycl -fsycl-targets=x86,fpga,gen9 -Xsycl-target=gen9 "-fmt:visa -fmt:spirv"`
+
+The `<arg>` parameter is passed by the driver directly to the SYCL device
+compiler for the corresponding target w/o parsing it. For each target there is
+some default code form which is generated in the absence of overriding via the
+`-Xsycl-target` option.
+
+**TBD:** Having multiple code forms for the same target in the fat binary might
+mean invoking device compiler multiple times. Multiple invocations are not
+needed if these forms can be dumped at various compilation stages by the single
+device compilation, like SPIR-V->visa->ISA. But if e.g. `gen9:visa3.2` and
+`gen9:visa3.3` are needed at the same time, then some mechanism is needed.
+Should it be a dedicated target triple for each needed visa version or Gen
+generation?
+
+#### Separate Compilation and Linking
+
+The compiler supports linking of device code obtained from different source
+files before generating the final SPIR-V to be fed to the back-end. The basic
+mechanism is to produce "fat objects" as a result of compilation - object files
```

Comment:
Is this some itemize list that was broken by some unexpected paragraph reformating in the source file?

Reply:
I don't think so. Could you clarify what formatting do you think would be better, please?